apps/studio/src/server/modules/user/user.service.ts

```diff
@@ -11,3 +21,144 @@ export const isUserDeleted = async (email: string) => {
 
   return user?.deletedAt ? true : false
 }
+
+export const validateEmailRoleCombination = ({
+  email,
+  role,
+}: {
+  email: string
+  role: ResourcePermission["role"]
+}) => {
+  if (!isGovEmail(email) && role === RoleType.Admin) {
+    throw new TRPCError({
+      code: "FORBIDDEN",
+      message:
+        "Non-gov.sg emails cannot be added as admin. Select another role.",
+    })
+  }
+}
+
+interface CreateUserProps {
+  email: User["email"]
+  name?: User["name"]
+  phone?: User["phone"]
+  role: ResourcePermission["role"]
+  siteId: ResourcePermission["siteId"]
+  trx?: SafeKysely // allows for transaction to be passed in from parent transaction
+}
+
+export const createUser = async ({
+  email,
+  name = "",
+  phone = "",
+  role = RoleType.Editor,
+  siteId,
+  trx,
+}: CreateUserProps) => {
+  if (!isEmail(email)) {
+    throw new TRPCError({
+      code: "BAD_REQUEST",
+      message: "Invalid email",
+    })
+  }
+
+  validateEmailRoleCombination({ email, role })
+
+  const isWhitelisted = await isEmailWhitelisted(email)
+  if (!isWhitelisted) {
+    throw new TRPCError({
+      code: "FORBIDDEN",
+      message: "There are non-gov.sg domains that need to be whitelisted.",
+    })
+  }
+
+  const executeInTransaction = async (tx: SafeKysely) => {
+    const user = await tx
+      .insertInto("User")
+      .values({
+        id: cuid2.createId(),
+        email,
+        name: name || email.split("@")[0] || "",
+        phone,
+      })
+      .onConflict((oc) =>
+        oc
+          .columns(["email", "deletedAt"])
+          .doUpdateSet((eb) => ({ email: eb.ref("excluded.email") })),
+      )
+      .returning(["id", "email", "name", "phone", "deletedAt"])
+      .executeTakeFirstOrThrow()
+
+    // unique constraint (@@unique([userId, siteId, resourceId, role]))
+    // does not work if one column is NULL e.g. resourceId
+    // Thus we have to check on the application layer
+    const existingResourcePermission = await tx
+      .selectFrom("ResourcePermission")
+      .where("userId", "=", user.id)
+      .where("siteId", "=", siteId)
+      .where("resourceId", "is", null) // because we are granting site-wide permissions
+      .selectAll()
+      .executeTakeFirst()
+
+    if (existingResourcePermission) {
+      throw new TRPCError({
+        code: "CONFLICT",
+        message: "User and permissions already exists",
+      })
+    }
```

Comment:
[question] why do we auto-resolve conflicting emails but not resource permissions? my take is that we should just automatically update it to the new requested permission because this is what the user wanted (ie, similar to updating a user's resource permission)

Reply:
removed this code in https://github.com/opengovsg/isomer/pull/1118/commits/ec9bdc45dcb5b851daff556ab200d8529abf5002

context: this check here was added prior to https://github.com/opengovsg/isomer/pull/1156 where the unique DB constraint do not work. Since it should be handled by the DB now, we can go ahead to remove it